### PR TITLE
Extended aggregation options in toolAggregate and calcOutput

### PR DESCRIPTION
**Pull request by tscheypidi**
This PR comes with a bunch of improvements in toolAggregate and calcOutput:

* `toolAggregate` now works also in cases in which a relation map `rel` is not specified but the information can be retrieved from the subdimensions of the object
* setNames are now properly set in `toolAggregate` in cases in which the number of subdimensions in reduced during aggregation (e.g. mapping all subdimensions in dim = 1 to only a single spatial dimension)
* `calcOutput` can now aggregate cellular data to regions via `aggregate = TRUE` if the data comes with a spatial "country" subdimension containing all iso countries
*`calcOutput` can now aggregate directly to the spatial subdimensions of the returned object (similar to `toolAggregate`)

### CODE REVIEW

File: R/toolAggregate.R [hunk 1]
```diff
@@ -83,13 +90,26 @@
...
+#'
 toolAggregate <- function(x, rel, weight = NULL, from = NULL, to = NULL, dim = 1, wdim = NULL, partrel = FALSE, # nolint
                           negative_weight = "warn", mixed_aggregation = FALSE, verbosity = 1) { # nolint
 
   if (!is.magpie(x)) stop("Input is not a MAgPIE object, x has to be a MAgPIE object!")
 
   comment <- getComment(x)
 
+  # create missing rel information from dimension names if argument "to" is set
+  if (missing(rel) && !is.null(to)) {
```
Comment (pfuehrlich-pik) [inline R/toolAggregate.R:102]: Can't we just add a default argument for `rel`? That makes it much clearer that it is actually allowed to not pass it.

File: tests/testthat/test-toolAggregate.R [hunk 2]
```diff
@@ -16,7 +16,7 @@ cfg <- getConfig(verbose = FALSE)
 
 noC <- function(x) {
   getComment(x) <- NULL
-  attr(x, "Metadata") <- NULL
+  attr(x, "Metadata") <- NULL # nolint: object_name_linter
```
Comment (pfuehrlich-pik) [inline tests/testthat/test-toolAggregate.R:19]:
```suggestion
  attr(x, "Metadata") <- NULL # nolint: object_name_linter.
```

File: R/calcOutput.R [hunk 3]
```diff
@@ -503,3 +404,139 @@ calcOutput <- function(type, aggregate = TRUE, file = NULL, years = NULL, # noli
...
+      # Use those columns only for aggregation that exist in either of the mappings
+      aggregate <- paste0(aggregateSplitted[commonColumns], collapse = "+")
+    }
+  }
+
+  items <- getItems(x, dim = 1)
+
+  # find mappings that have the same (setequal) items (usually ISO countries) in any column as data (items)
+  # lapply loops over the mappings, vapply loops over the columns of a mapping
+  columnHasItems <- lapply(rel, vapply, setequal, items, FUN.VALUE = logical(1))
```
[the remark this answers is not in the document]
Comment (tscheypidi) [reply R/calcOutput.R:498]: `setequal`  is what we want here

File: R/calcOutput.R [hunk 4]
```diff
@@ -503,3 +404,139 @@ calcOutput <- function(type, aggregate = TRUE, file = NULL, years = NULL, # noli
...
+    rel <- rel[[1]]
+  }
+
+  # If there are more than one fitting mappings merge them. If column names from the first mapping (given via
+  # 'regionmapping') also exist in further mappings (provided via 'extramappings') keep only the columns from
+  # the first mapping
+  if (length(relFitting) > 1) {
+    itemCol <- columnNameWithItems[relFitting]
+    tmp <- rel[[1]]
+    for (i in 2:length(rel)) {
```
[the remark this answers is not in the document]
Comment (tscheypidi) [reply R/calcOutput.R:529]: the code before ensures that rel is >= 2 as otherwise relFitting could not be >=2

File: R/toolAggregate.R (hunk 1 above)
Comment (tscheypidi) [reply R/toolAggregate.R:102]: I thought about it, but to me it looks more secure to do it this way. Alternative would be to set `rel = NULL` but I suspect that there might be cases where the provided mapping by accident is `NULL`. This cases should not trigger the automatic generation of a mapping.

File: R/calcOutput.R [hunk 5]
```diff
@@ -364,80 +313,33 @@ calcOutput <- function(type, aggregate = TRUE, file = NULL, years = NULL, # noli
...
+    # to be used in the aggregation function.
+    if (is.null(x$aggregationArguments)) x$aggregationArguments <- list()
+    if (!is.list(x$aggregationArguments)) stop("x$aggregationArguments must be a list of function arguments!")
+    # Add base arguments to the argument list (except of rel, which is added later)
+    x$aggregationArguments$x <- quote(x$x)
+    if (!is.null(x$weight)) x$aggregationArguments$weight <- quote(x$weight)
+    if (x$mixed_aggregation) x$aggregationArguments$mixed_aggregation <- TRUE # nolint
+
+    x$aggregationArguments$rel <- quote(map$rel)
+    if (aggregate != TRUE) x$aggregationArguments$to <- map$aggregate
```
[the remark this answers is not in the document]
Comment (tscheypidi) [reply R/calcOutput.R:342]: aggregate is not alway boolean, hence the check should be instead:

```suggestion
    if (!isTRUE(aggregate) )  x$aggregationArguments$to <- map$aggregate
```

**Review by pfuehrlich-pik — APPROVED** [review]

**Pull request merged**